token/program-2022/src/extension/token_group/processor.rs

```diff
@@ -0,0 +1,107 @@
+        msg,
+        program_error::ProgramError,
+        program_option::COption,
+        pubkey::Pubkey,
+    },
+    spl_token_group_interface::{
+        error::TokenGroupError,
+        instruction::{InitializeGroup, TokenGroupInstruction},
+        state::TokenGroup,
+    },
+};
+
+fn realloc_mint(mint_info: &AccountInfo, extension: ExtensionType) -> Result<(), ProgramError> {
+    let extension_len = extension.try_get_tlv_len()?;
+    let new_account_len = mint_info
+        .data_len()
+        .checked_add(extension_len)
+        .ok_or::<ProgramError>(TokenError::Overflow.into())?;
+    mint_info.realloc(new_account_len, false)?;
+    Ok(())
+}
+
+/// Processes a [InitializeGroup](enum.TokenGroupInstruction.html) instruction.
+pub fn process_initialize_group(
+    _program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    data: InitializeGroup,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let group_info = next_account_info(account_info_iter)?;
+    let mint_info = next_account_info(account_info_iter)?;
+    let mint_authority_info = next_account_info(account_info_iter)?;
+
+    // check that the mint and group accounts are the same, since the group
+    // extension should only describe itself
+    if group_info.key != mint_info.key {
+        msg!("Group configurations for a mint must be initialized in the mint itself.");
+        return Err(TokenError::MintMismatch.into());
+    }
+
+    // scope the mint authority check, since the mint is in the same account!
+    {
+        // This check isn't really needed since we'll be writing into the account,
+        // but auditors like it
+        check_program_account(mint_info.owner)?;
+        let mint_data = mint_info.try_borrow_data()?;
+        let mint = StateWithExtensions::<Mint>::unpack(&mint_data)?;
+
+        if !mint_authority_info.is_signer {
+            return Err(ProgramError::MissingRequiredSignature);
+        }
+        if mint.base.mint_authority.as_ref() != COption::Some(mint_authority_info.key) {
+            return Err(TokenGroupError::IncorrectMintAuthority.into());
+        }
+
+        if mint.get_extension::<GroupPointer>().is_err() {
+            msg!(
+                "A mint with group configurations must have the group-pointer extension \
+                 initialized"
+            );
+            return Err(TokenError::InvalidExtensionCombination.into());
+        }
+    }
+
+    // Reallocate the mint for the new extension
+    realloc_mint(mint_info, ExtensionType::TokenGroup)?;
+
+    // Allocate a TLV entry for the space and write it in
+    // Assumes that there's enough SOL for the new rent-exemption
+    let mut mint_data = mint_info.try_borrow_mut_data()?;
+    let mut mint = StateWithExtensionsMut::<Mint>::unpack(&mut mint_data)?;
+    let group = mint.init_extension::<TokenGroup>(false)?;
+    *group = TokenGroup::new(mint_info.key, data.update_authority, data.max_size.into());
```

Comment:
@joncinque I'm currently debugging why this bit of code doesn't work properly only _some of the time_. I'm wondering if it's got anything to do with the `realloc` involved here.

I'm expecting `init_extension(..)` to fail if the extension exists, and most of the time it does, but every so often (particularly in CI) it doesn't fail, and my test expecting `ExtensionAlreadyInitialized` fails.

Any ideas lmk.